LongestPole/Main.lean

```diff
@@ -0,0 +1,191 @@
+/-
+Copyright (c) 2023 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import ImportGraph
+import Mathlib.Data.String.Defs
+import Std.Lean.Util.Path
+import Cli
+
+/-!
+# `lake exe pole`
+
+Longest pole analysis for Mathlib build times.
+-/
+
+set_option autoImplicit true
+
+open Cli
+
+open Lean Meta
+
+open Lean Core System
+
+/-- Runs a terminal command and retrieves its output -/
+def runCmd (cmd : String) (args : Array String) (throwFailure := true) : IO String := do
+  let out ← IO.Process.output { cmd := cmd, args := args }
+  if out.exitCode != 0 && throwFailure then throw $ IO.userError out.stderr
+  else return out.stdout
+
+def runCurl (args : Array String) (throwFailure := true) : IO String := do
+  runCmd "curl" args throwFailure
+
+def speedCenterRunJson (hash : String) : IO String :=
+  -- e7b27246-a3e6-496a-b552-ff4b45c7236e is the `repo_id` for the `mathlib4` repository.
+  runCurl #["http://speed.lean-fro.org/mathlib4/api/run/e7b27246-a3e6-496a-b552-ff4b45c7236e?hash=" ++ hash]
+
+structure SpeedCenterSource_1 where
+  repo_id : String
+  hash : String
+deriving ToJson, FromJson
+
+structure SpeedCenterSource_2 where
+  source : SpeedCenterSource_1
+deriving ToJson, FromJson
```

Comment:
What is the use of these two structures? Is this to get the Lean structures aligned with some JSON data structure?

Reply:
Yes. See for example this API output http://speed.lean-fro.org/mathlib4/api/run/e7b27246-a3e6-496a-b552-ff4b45c7236e?hash=9aded8e01ba2ac52c5e6864d9847cf26c6aaa015

Reply:
Can you put these all in a `SpeedCenterApi` namespace to it this clear that these follow that json format, and should be kept in sync with the API? You could then drop the `SpeedCenter` prefices.